src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs

```diff
@@ -13,869 +13,614 @@
-                if (_extendedProtectionPolicy.PolicyEnforcement == PolicyEnforcement.WhenSupported)
-                {
-                    flags |= Interop.SspiCli.ContextFlags.AllowMissingBindings;
-                }
+                int maxCount = checked(Int32.MaxValue - 4 - sizes.BlockSize - sizes.SecurityTrailer);
 
-                if (_extendedProtectionPolicy.PolicyEnforcement != PolicyEnforcement.Never &&
-                    _extendedProtectionPolicy.ProtectionScenario == ProtectionScenario.TrustedProxy)
+                if (count > maxCount || count < 0)
                 {
-                    flags |= Interop.SspiCli.ContextFlags.ProxyBindings;
+                    throw new ArgumentOutOfRangeException("count", SR.Format(SR.net_io_out_range, maxCount));
                 }
             }
-            else
+            catch (Exception e)
             {
-                // Server side should not request any of these flags.
-                if (protectionLevel != ProtectionLevel.None)
+                if (!ExceptionCheck.IsFatal(e) && GlobalLog.IsEnabled)
                 {
-                    flags |= Interop.SspiCli.ContextFlags.MutualAuth;
+                    GlobalLog.Assert("NTAuthentication#" + LoggingHash.HashString(securityContext) + "::Encrypt", "Arguments out of range.");
                 }
 
-                if (impersonationLevel == TokenImpersonationLevel.Identification)
-                {
-                    flags |= Interop.SspiCli.ContextFlags.InitIdentify;
-                }
-
-                if (impersonationLevel == TokenImpersonationLevel.Delegation)
-                {
-                    flags |= Interop.SspiCli.ContextFlags.Delegate;
-                }
-            }
-
-            _canRetryAuthentication = false;
-
-            try
-            {
-                _context = new NTAuthentication(isServer, package, credential, servicePrincipalName, flags, channelBinding);
-            }
-            catch (Win32Exception e)
-            {
-                throw new AuthenticationException(SR.net_auth_SSPI, e);
+                throw;
             }
-        }
 
-        private Exception SetException(Exception e)
-        {
-            if (_exception == null || !(_exception is ObjectDisposedException))
+            int resultSize = count + sizes.SecurityTrailer + sizes.BlockSize;
+            if (output == null || output.Length < resultSize + 4)
             {
-                _exception = e;
+                output = new byte[resultSize + 4];
             }
 
-            if (_exception != null && _context != null)
-            {
-                _context.CloseContext();
-            }
+            // Make a copy of user data for in-place encryption.
+            Buffer.BlockCopy(buffer, offset, output, 4 + sizes.SecurityTrailer, count);
 
-            return _exception;
-        }
+            // Prepare buffers TOKEN(signature), DATA and Padding.
+            var securityBuffer = new SecurityBuffer[3];
```

Comment:
Rather than potentially having errors in the indices (and length), consider:

```C#
SecurityBuffer[] securityBuffer =
{
    new SecurityBuffer(output, 4, sizes.SecurityTrailer, SecurityBufferType.Token),
    new SecurityBuffer(output, 4 + sizes.SecurityTrailer, count, SecurityBufferType.Data),
    new SecurityBuffer(output, 4 + sizes.SecurityTrailer + count, sizes.BlockSize, SecurityBufferType.Padding),
};
```